routing/bicycle_directions.cpp

```diff
@@ -0,0 +1,185 @@
+  {
+    return m_pathSegments;
+  }
+
+  virtual void GetPossibleTurns(TNodeId node, m2::PointD const & ingoingPoint,
+                                m2::PointD const & junctionPoint,
+                                size_t & ingoingCount,
+                                TTurnCandidates & outgoingTurns) const override
+  {
+    if (node >= m_routeEdges.size() || node >= m_adjacentEdges.size())
+    {
+      ASSERT(false, ());
+      return;
+    }
+
+    AdjacentEdgesMap::const_iterator adjacentEdges = m_adjacentEdges.find(node);
+    if (adjacentEdges == m_adjacentEdges.cend())
+    {
+      ASSERT(false, ());
+      return;
+    }
+
+    ingoingCount = adjacentEdges->second.ingoingTurnCount;
+    outgoingTurns = adjacentEdges->second.outgoingTurns;
+  }
+
+  virtual double GetPathLength() const override { return m_routeLength; }
+
+  virtual m2::PointD const & GetStartPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.front().GetStartJunction().GetPoint();
+  }
+
+  virtual m2::PointD const & GetEndPoint() const override
+  {
+    CHECK(!m_routeEdges.empty(), ());
+    return m_routeEdges.back().GetEndJunction().GetPoint();
+  }
+
+private:
+   IRoadGraph::TEdgeVector const & m_routeEdges;
+   AdjacentEdgesMap const & m_adjacentEdges;
+   TUnpackedPathSegments const & m_pathSegments;
+   double m_routeLength;
+};
+
+ftypes::HighwayClass GetHighwayClass(FeatureID const & featureId, Index const & index)
+{
+  ftypes::HighwayClass highWayClass = ftypes::HighwayClass::Undefined;
+  MwmSet::MwmId const & mwmId = featureId.m_mwmId;
+  uint32_t const featureIndex = featureId.m_index;
+
+  FeatureType ft;
+  Index::FeaturesLoaderGuard loader(index, mwmId);
+  loader.GetFeatureByIndex(featureIndex, ft);
+  ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
+  highWayClass = ftypes::GetHighwayClass(ft);
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Error, ());
+  ASSERT_NOT_EQUAL(highWayClass, ftypes::HighwayClass::Undefined, ());
+  return highWayClass;
+}
+}  // namespace
+
+namespace routing
+{
+BicycleDirectionsEngine::BicycleDirectionsEngine(Index const & index) : m_index(index)
+{
+}
+
+void BicycleDirectionsEngine::Generate(IRoadGraph const & graph, vector<Junction> const & path,
+                                       Route::TTimes & times,
+                                       Route::TTurns & turnsDir,
+                                       vector<m2::PointD> & routeGeometry,
+                                       my::Cancellable const & cancellable)
+{
+  LOG(LINFO, ("BicycleDirectionsEngine::Generate"));
+  CHECK_GREATER(path.size(), 1, ());
+  times.clear();
+  turnsDir.clear();
+  routeGeometry.clear();
+  m_adjacentEdges.clear();
+  m_pathSegments.clear();
+
+  CalculateTimes(graph, path, times);
+
+  auto emptyPathWorkaround = [&]()
+  {
+    turnsDir.emplace_back(path.size() - 1, turns::TurnDirection::ReachedYourDestination);
```

Comment:
|path| is passed to BicycleDirectionsEngine::Generate() by a constant ref (vector<Junction> const & path). So it cannot be modified by in ReconstructPath. On the other hand the code is changed after some of your former comments. Not it looks like:
```c++
void BicycleDirectionsEngine::Generate(IRoadGraph const & graph, vector<Junction> const & path,
                                       Route::TTimes & times,
                                       Route::TTurns & turns,
                                       vector<m2::PointD> & routeGeometry,
                                       my::Cancellable const & cancellable)
{
  size_t const pathSize = path.size();
  CHECK_NOT_EQUAL(pathSize, 0, ());

  times.clear();
  turns.clear();
  routeGeometry.clear();
  m_adjacentEdges.clear();
  m_pathSegments.clear();

  auto emptyPathWorkaround = [&]()
  {
    turns.emplace_back(pathSize - 1, turns::TurnDirection::ReachedYourDestination);
    this->m_adjacentEdges[0] = AdjacentEdges(1); // There's one ingoing edge to the finish.
  };

  if (pathSize <= 1)
  {
    ASSERT(false, (pathSize));
    emptyPathWorkaround();
    return;
  }

  CalculateTimes(graph, path, times);

  IRoadGraph::TEdgeVector routeEdges;
  if (!ReconstructPath(graph, path, routeEdges, cancellable))
  {
    LOG(LDEBUG, ("Couldn't reconstruct path"));
    emptyPathWorkaround();
    return;
  }
  if (routeEdges.empty())
  {
    ASSERT(false, ());
    emptyPathWorkaround();
    return;
  }
....
```
As you see there's a check that |path| is not empty and it cannot be changed.